torch/csrc/distributed/c10d/ProcessGroupGloo.cpp

```diff
@@ -465,9 +469,37 @@ ProcessGroupGloo::AsyncWork::AsyncWork(
     std::vector<std::vector<at::Tensor>> outputTensors,
     const char* profilingTitle,
     const c10::optional<std::vector<at::Tensor>>& inputTensors)
-    : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
+    // Specify nullptr to profilingTitle of Parent Class is to disable the profiling in Parent Class.
+    // It's because general ProcessGroup profiling is kicked off upon construction of Work and it's not
+    // reasonable for async work. We will add a new record function in this AsyncWork and delay the profiling
+    // start time to when the the work really executed, not constructed.
+    : ProcessGroup::Work(-1, OpType::UNKNOWN, nullptr, inputTensors),
       outputTensors_(std::move(outputTensors)),
-      future_(createFutureAsOutput(outputTensors)) {}
+      future_(createFutureAsOutput(outputTensors)) {
+        if (profilingTitle != nullptr) {
+            auto recordingFunction =
+                std::make_shared<at::RecordFunction>(at::RecordScope::USER_SCOPE);
+            std::function<void()> before_handler = [this, inputTensors, profilingTitle, recordingFunction]() {
+              if (recordingFunction->isActive()) {
+                // This Work will be start and completed by another thread.
+                recordingFunction->_setAsync();
+                std::vector<c10::IValue> inputs;
+                if (inputTensors) {
+                  inputs.reserve(inputTensors->size());
+                  for (const auto& tensor : *inputTensors) {
+                    inputs.push_back(tensor);
+                  }
+                }
+                recordingFunction->before(profilingTitle, inputs);
+              }
+            };
+            recordFunctionBeforeCallback_ = at::wrapPropagateTLSState(before_handler);
+            std::function<void()> end_handler = [this, recordingFunction]() {
+              recordingFunction->end();
+            };
+            recordFunctionEndCallback_ = at::wrapPropagateTLSState(end_handler);
+        }
+      }
```

Comment:
Is the clang lint error here?

Reply:
It seems not. I fixed it and the clang also fail.